### PR TITLE
Add getNewShuffling helper function

### PR DESCRIPTION
**Pull request by wemeetagain**
- Use `Buffer` for `hash32` type - Previously, we'd defined `hash32` to be of type `number`. I'm assuming that was just a stand-in. Here we use a `Buffer`, so we can reuse the `ethereumjs-util` packaged code.
- Change the typescript compiler built-in type declarations to include newer APIs - useful especially for allowing newer `Array` methods (eg: `find`, `includes`, etc.)
- Add `get_new_shuffling` utility function from the [spec](https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#get_new_shuffling)
- Add basic sanity test

### CODE REVIEW

**Comment by GregTheGreek** [comment]
@wemeetagain thanks for the contribution! Going over this shortly.

**Comment by GregTheGreek** [comment]
@wemeetagain Awesome work! I've been working on the `shuffle()` (93ed0b1da600b9e82a484fdce46df7767ce5e94a) for a while but ran into the whole bytes/buffer dilemma (#59 & #26). Internally (we're trying to be more "public" with our discussions) we've been debating if we want to use browserify/webpack to facilitate node packages, or go the pure vanilla JS to maintain browser compatibility...

I'm going to chat with @ansermino about this (feel free to join, we can setup a call), because `Buffers` were in the original implementation (pre-typescript) and we opted to remove them. Although, using buffers will make things much easier.

File: beaconChain/helpers/stateTransitionHelpers.ts [hunk 1]
```diff
@@ -38,22 +39,46 @@ function shuffle<T>(values: T[], seed: hash32): T[] {
...
   // It is a logic error to supply an oversized list.
-  if (valuesCount < randMax) { throw new Error("Oversized list supplied to shuffle()!"); }
+  if (!(valuesCount < randMax)) { throw new Error("Oversized list supplied to shuffle!"); }
 
   // Make a copy of the values
   const output: T[] = values.slice();
-  const source = seed; // REALLY??
-  const index = 0; // REALLY??
+  let source = toBuffer(seed);
+  let index = 0;
```
Comment (GregTheGreek) [inline beaconChain/helpers/stateTransitionHelpers.ts]: Please add the types for 46 & 47

File: beaconChain/helpers/stateTransitionHelpers.ts [hunk 2]
```diff
@@ -38,22 +39,46 @@ function shuffle<T>(values: T[], seed: hash32): T[] {
...
   while (index < valuesCount - 1) {
     // Re-hash the `source` to obtain a new pattern of bytes.
-    // TODO figure out what this hash function is in python -> JS
-    // let source = hash(source)
+    source = keccak256(source).slice(0, 32);
```
Comment (GregTheGreek) [inline beaconChain/helpers/stateTransitionHelpers.ts]: here as well :)

**Review by GregTheGreek — CHANGES_REQUESTED** [review]
Could you please add types to all the constants inside of `shuffle()` :)

File: beaconChain/helpers/stateTransitionHelpers.ts (hunk 2 above)
Comment (wemeetagain) [reply beaconChain/helpers/stateTransitionHelpers.ts]: Not able to set the type here as its already been declared.

**Comment by wemeetagain** [comment]
@GregTheGreek Thanks for highlighting that bytes/buffer dilemma.
It would be great to have a development environment that allows us to write modern typescript (with es6/7/etc. language features) without worrying about browser API compatability differences.

File: beaconChain/helpers/stateTransitionHelpers.ts (hunk 2 above)
Comment (GregTheGreek) [reply beaconChain/helpers/stateTransitionHelpers.ts]: cool makes sense!

**Comment by noot** [comment]
@wemeetagain thanks for the PR! I've discussed with @GregTheGreek and we think the best course of action is to represent the `hash32` type as a hex-formatted string eg. `seed = 0x1234abcd` 
This way, we can use `sample_from_source = parseInt(seed.slice(pos, pos+6), 16)` to grab 3 bytes of the string and parse it as base16 to an int. I use `seed.slice(pos, pos+6)` since each hex character represents half a byte, and we need 3 bytes of the seed for each iteration per the spec, thus 6 characters total.

Let me know if this makes sense! I can write this out in code form soon.

**Comment by wemeetagain** [comment]
@noot I understand. I can't say that it feels particularly good though. This has pretty wide-reaching consequences. (This is not the only place where hashes will be used, obviously)

Why not use the native byte array representation? Why not use the same type that most crypto libraries will give you? At the end of the day, our hash32 is meant to be an array of inspectable bytes, so it makes more sense to use that directly if available than to shoehorn a string to do the same.
Ie. its simpler to deal with a byte array directly than a utf-8-encoded, hex-encoded byte array (with two additional characters '0x' prepended to it).

Of course, we can simply convert all Buffers to strings, chop off the '0x' where necessary, but these extra conversions and mental gymnastics are not necessary if we just use a representation that fits the bill. My opinion is that over time, these unnecessary conversions cloud the waters of "what's actually happening" in the code and will not be worth the perceived short-term convenience.

I'm curious if yall have more insight into your conversation.

**Comment by JonathanLorimer** [comment]
@noot @GregTheGreek 

I had a similar idea to @wemeetagain. uint8array - uint32array (native javascript) might provide us the functionality we need. However, I was thinking that we might create a library or a utility that converts these numbers to arrays, if its a hex representation it could be an array of double characters or an array of tuples (arrays with length 2), if binary just an array of 1s and 0s. This provides a lot of flexibility, as the Array prototype is fairly robust and has good methods.

It makes things like shuffling values at specific indices really easy!

**Comment by GregTheGreek** [comment]
@wemeetagain @noot @JonathanLorimer 

@wemeetagain I totally agree, that does make sense. @JonathanLorimer brought that up to me earlier. Would you be interested in implementing it for `shuffle()`? @JonathanLorimer I actually think it might be a good use case to isolated it in its own library. Ultimately we want to create as many useful libraries for JS ecosystem as possible, this definitely fits the mold! @wemeetagain Thoughts?

side note: 
We're planning for a call on friday to disucss roadmap etc... Would love for all of you to join, I'll post a link in the gitter tomorrow.

**Comment by wemeetagain** [comment]
@GregTheGreek Yeah I can take a crack at it.

Which functionality are you thinking to break out into its own library?

Definitely interested in joining Friday. :smiley_cat:

**Comment by GregTheGreek** [comment]
@wemeetagain depending on what it looks like (maybe it's not a lot of functionality) but the uint8array as a hash32. I'm thinking we could extend it further to support generic lengths etc... That way more people can use it!

I'm a big fan of taking custom implementation related code into small libraries, because there's definitely someone out there that may benefit from it.

That being said, perhaps it's overkill, but worth exploring. 

I'm interested in helping you out I know Jonathan is also interested.

**Comment by JonathanLorimer** [comment]
It seems that there is a lot of interest in building this library. May I suggest, on the principle of division of labour, that one of you (@wemeetagain and @GregTheGreek) work on the implementation of shuffle() and the other works on the data structure/library, I can take care of specing the the requirements with tests and then follow up with documentation on the API.

also a note: uint32array and int32arrays were introduced as of ES2017 so there might not even be a need for a library after all

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Int32Array
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint32Array

If I understand correctly, the Hash32 is a hex string though, so we might still need to implement a 32BitHexArray Type/Construct

**Comment by GregTheGreek** [comment]
Im on mobile, so I'll have a look in a bit. But in regards to the library perhaps let's get it working in shuffle and then let's seperate it if it makes sense

**Comment by ChainSafeSystems** [comment]
just created a [repo](https://github.com/ChainSafeSystems/byte-array-js) for this to live in :)

**Comment by noot** [comment]
@GregTheGreek @wemeetagain @JonathanLorimer yeah I think an array of hex characters makes sense; it looks like uint8 array is basically the same thing though. I guess the main difference between uint8 array and our hash32 array is that the hash32 array elements are hex-encoded, does it make sense to make a library for this or would a few helper functions be ok? let me know what you guys think.

**Comment by JonathanLorimer** [comment]
@noot @wemeetagain @GregTheGreek 

Normally I would advocate for just a few helper functions in the interest of simplicity. However, as I understand it, the ethos of this project is to create some libraries for the ecosystem. Additionally it might be useful to make a hash32 type, and I think abstracting this interface into a library might prove useful. 

A final consideration is that the Beacon spec never (as far as I can see) defines the Hash32 type, it just makes use of it, so in the interest of keeping the Lodestar implementation as close to the spec as possible I think it makes sense to abstract the typing out into a separate repo.

But again, the helper functions would be probably get things up and running quicker.

**Comment by GregTheGreek** [comment]
Community consensus is fine by me.
Personally @wemeetagain if you want to try and implement it directly in the code, we can always move it out the the [repo that aidan create](https://github.com/ChainSafeSystems/byte-array-js) if it makes sense. (name is stubbed)

**Comment by wemeetagain** [comment]
I've been doing some digging into the built-in Uint8Array and possible directions for us:

The [Uint8Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array) is built into javascript directly, but its based on TypedArray, vs Array, and so has a slightly more limited standard api.

Luckily, the polkadotjs team has a pretty great [utility library](https://github.com/polkadot-js/common/tree/master/packages/util/src) that includes a lot of functions that operate on/convert between Uint8Arrays, Buffers, numbers, hex strings, bignumbers, etc.

We primarily need the following operations:
- get/set by index
- slice array into subarray
- concat multiple arrays
- convert array to bignumber, hex string, etc
- convert subarray to uint of variable length
- (missing anything else?)

All of these operations are supported via the standard api and/or the polkadot utility library, EXCEPT for converting a subarray to uint (like in [buffer.readUIntBE](https://nodejs.org/api/buffer.html#buffer_buf_readuintbe_offset_bytelength)) (The built-in [DataView](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView#Read) methods come close, but no variable length read/write exists, also no big/little endian toggle).

This last requisite function/method could be pulled into its own library, etc if we decide to go this route. (It's implemented in the buffer library referenced lower.)

Alternatively, we could use the [buffer](https://github.com/feross/buffer) library, and use the Buffer object, which shares the api of the node standard library's [Buffer](https://nodejs.org/api/buffer.html), which is more flexible and includes a few more of our desired operations natively.

From the README:

> The Buffer constructor returns instances of Uint8Array that have their prototype changed to Buffer.prototype. Furthermore, Buffer is a subclass of Uint8Array, so the returned instances will have all the node Buffer methods and the Uint8Array methods. Square bracket notation works as expected -- it returns a single octet.

We've been hesitant to use the node Buffer api because we want to support web usage natively, and see using a node api as antithetical to that end. However, in this case we _could_ use the slightly more featureful Buffer api and still retain full web support.

TL;DR
We can use Uint8Array + utility functions or Buffer. I'm hoping that these notes can inform discussion and help us move in a certain direction.

edit: I didn't refresh this PR and didn't see the recent (1dayish old) comments. Will update in a bit.

**Comment by JonathanLorimer** [comment]
@wemeetagain we could put the uint8array to uint into the fixed sized number library we already built

**Comment by wemeetagain** [comment]
Reading through everyone's comments, it looks like consensus is starting to converge around the use of Uint8Array as the representation for byte arrays.

I'm curious to hear more thoughts about the use of feross/buffer's Buffer in light of my previous comment, but if discussion has died down, then:

I'll update this PR to use Uint8Array instead of Buffer, including the `readUintBE` utility function in line (which will be copy/pasted from the feross/buffer library).

Once this is landed, we can figure out the proper library-ification.

**Comment by GregTheGreek** [comment]
@wemeetagain [client side buffer](https://github.com/feross/buffer) is interesting. I'm doing more research on this. I'm fine with it being done with uint8arrray and we can always optimize it later if we find it to be unsuitable.

At first glance though it seems most operations are slower (based on the benchmarks) for the client side buffer.

**Comment by wemeetagain** [comment]
@GregTheGreek yeah, I was liking Buffer mainly for its ubiquitous use in the js ecosystem and its featureful api. in other words, familiarity and convenience.

Ultimately, neither of these are probably good enough reasons to move in that direction though.
The familiarity and ubiquity of Buffer is mostly a result of native typed arrays not existing until relatively recently.
And the convenience of the api can be recreated/emulated with utility libraries.

**Comment by GregTheGreek** [comment]
I'm on mobile, but I've gone through the code looks good so far. I want to go through the polkadot library quickly when I get home

File: beaconChain/helpers/stateTransitionHelpers.ts [hunk 3]
```diff
@@ -22,38 +23,72 @@ export function getActiveValidatorIndices(validators: ValidatorRecord[]): int[]
...
-  const randMax = 2 ** (randBytes * 8) - 1;
+  const randMax: number = 2 ** (randBytes * 8) - 1;
 
   // The range of the RNG places an upper-bound on the size of the list that may be shuffled.
   // It is a logic error to supply an oversized list.
-  if (valuesCount < randMax) { throw new Error("Oversized list supplied to shuffle()!"); }
+  if (!(valuesCount < randMax)) { throw new Error("Oversized list supplied to shuffle!"); }
 
   // Make a copy of the values
   const output: T[] = values.slice();
```
Comment (ansermino) [inline beaconChain/helpers/stateTransitionHelpers.ts:55]: Any foreseeable issues with using shallow copy?

File: beaconChain/helpers/stateTransitionHelpers.ts [hunk 4]
```diff
@@ -22,38 +23,72 @@ export function getActiveValidatorIndices(validators: ValidatorRecord[]): int[]
...
-  const source = seed; // REALLY??
-  const index = 0; // REALLY??
+  let source: Uint8Array = seed;
+  let index: number = 0;
   while (index < valuesCount - 1) {
     // Re-hash the `source` to obtain a new pattern of bytes.
-    // TODO figure out what this hash function is in python -> JS
-    // let source = hash(source)
+    source = keccakAsU8a(source).slice(0, 32);
```
Comment (ansermino) [inline beaconChain/helpers/stateTransitionHelpers.ts]: Is this `slice()` necessary? keccakAsU8a uses keccak256 which always produces 32bytes, passing in a `source` of more than 32 elements (bytes) also returns 32 elements (bytes)

File: beaconChain/helpers/stateTransitionHelpers.ts [hunk 5]
```diff
@@ -22,38 +23,72 @@ export function getActiveValidatorIndices(validators: ValidatorRecord[]): int[]
   }, []);
 }
 
+// Modified from: https://github.com/feross/buffer/blob/master/index.js#L1125
+function readUIntBE(array: Uint8Array, offset: number, byteLength: number): number {
```
Comment (ansermino) [inline beaconChain/helpers/stateTransitionHelpers.ts]: Are we planning on using this elsewhere? If so, perhaps some tests specifically for this would be beneficial

**Review by ansermino — COMMENTED** [review]
Just some observations, awesome work Cayman!

File: beaconChain/helpers/stateTransitionHelpers.ts (hunk 4 above)
Comment (wemeetagain) [reply beaconChain/helpers/stateTransitionHelpers.ts]: No, the slice isn't necessary. I'll just add a comment referring to the assumed length.

File: beaconChain/helpers/stateTransitionHelpers.ts (hunk 3 above)
Comment (wemeetagain) [reply beaconChain/helpers/stateTransitionHelpers.ts:55]: I don't think so. Do you think it could be an issue?

File: beaconChain/helpers/stateTransitionHelpers.ts (hunk 5 above)
Comment (wemeetagain) [reply beaconChain/helpers/stateTransitionHelpers.ts]: We'll probably use this elsewhere?

Since I copied this code from feross/buffer, I'll just copy over the relevant tests, too. :)

I'm assuming this will get pulled into a utility library at some point eventually/soonish. That might be a good time to revisit its testing robustness.

File: beaconChain/helpers/stateTransitionHelpers.ts (hunk 3 above)
Comment (ansermino) [reply beaconChain/helpers/stateTransitionHelpers.ts:55]: Probably not. Its a generic function so it might be naive to expect anything of the inputs, but all of the objects we have defined so far are shallow.

File: beaconChain/helpers/stateTransitionHelpers.ts (hunk 5 above)
Comment (ansermino) [reply beaconChain/helpers/stateTransitionHelpers.ts]: Awesome, thank you :) 

I was thinking the same thing, in fact the ssz-js library has some similar functionality, although it interfaces quite differently (https://github.com/ChainSafeSystems/ssz-js/blob/master/src/intBytes.js). Lets keep an eye out for more similarities and migrate when appropriate.

**Review by ansermino — APPROVED** [review]

**Comment by ansermino** [comment]
@GregTheGreek please close

**Review by GregTheGreek — APPROVED** [review]
